### ISSUE TITLE
Customising "spaces" in search queries.

### ISSUE THREAD

**Issue by mrshanepaul** [CanisLupus/swift-selection-search#85]
Firstly, what an awesome plugin. Excellent work. better than others I have tried. :)
My issue:
Some sites do not use the default spaces for search terms. I have looked through the docs and other issues but cannot find a way to customise them? 
I surely cannot be the first person to have this issue?

Example 1: "my: search term "
Setting:    https://domain.com/search?f={searchTerms}
Wanted:   https://domain.com/search?f=my+search+term
Actual:     https://domain.com/search?f=my%3A%20search%20term%20
Result:     Website fails to load

Example 2: "my: search term "
Setting:    https://domain.com/search/={searchTerms}
Wanted:   https://domain.com/search/my-search-term
Actual:     https://domain.com/search?f=my search term
Result:     Some sites can handle this.

I apologise if there is an option for this already. If there is, my suggestion is to add help/info about this on the settings page. 😃

**Comment by ZIEXED** [comment]
This is a different problem but since it can be easily fixed and related to Spaces I thought it won't be a problem to just report it here... 
We need to trim selected characters before searching, most sites can handle it correctly but I found some sites where search results affected by the additional spaces.

**Comment by CanisLupus** [comment]
Hey everyone!

To be honest, I hadn't faced that problem myself with the engines I use, so I didn't even think this was an option SSS needed. I guess that says "no, there's no option for this". :) It could be solved by some configuring for {searchTerms}, like using {searchTerms:( ,+)(_,-)} to specify pairs of replacements. In this case, spaces should be replaced by "+", and underscores by "-" (just examples).

One problem I'm beginning to have is explaining these small things to people that are not as tech-savvy. I'm also getting a pretty long backlog of suggestions to implement, but this one is definitely on the more important side. :)

Can I please have an example of a real search URL where SSS doesn't work correctly due to spaces? It's much appreciated!

Cheers for the suggestion and detailed explanation!
Daniel

**Comment by mrshanepaul** [comment]
An example is very easy: wikipeadia.
Search with default wiki search:
    "my: search term "
Result:
https://my.wikipedia.org/wiki/Search_term
And in a language I don't recognise...! 😁 

I think the colon is causing the trouble. Hard to say what other chars would. 

The sites seem to handle spaces rather than their default now I am testing it more. It was the colon causing trouble. Needs more fibre...

It might not be worth adding replacements if this is the case or at least putting it on the back-burner. :)

I cross checked it with another chrome addon called "Selection Search" I still had installed from when I used chrome. They do not have an option to replace chars so I am assuming from that it is not needed.

I think they just replace special chars like ':' with a space and then perhaps nuke additional spaces.

A line of regex to fix methinks? 😃

**Comment by CanisLupus** [comment]
@mrshanepaul I see. Interesting behaviour from Wikipedia. The Internet tells me that "my" is the country code for Myanmar. It might be assuming the language from the colon, yup. Your list of tests will be useful when I implement something to fix this. Thanks! ;)

I can't assume that special characters should all be stripped since I can think of at least two sites that search for exact text, which are www.gamerankings.com and symbolhound.com. GameRankings needs a colon to be there if the game has a colon in the name, and SymbolHound kind of... _depends_ on it. ;) That said a checkbox to disable the stripping would solve it for both these cases. But maybe there are more specific instances of this problem... hmm...

@ZIEXED Sorry I wasn't clear! I meant that as an example of what could solve the problem. It isn't implemented. Re-reading, I think my wording was not obvious. :)

**Comment by mrshanepaul** [comment]
Myanmar. Of course! I just failed geography. :) The language was Burmese and I have not seen that before.
I agree. I would also not have the checkbox on by default for sites as most sites seem to handle it fine.

**Comment by mrshanepaul** [comment]
Just a tip while I was mucking about with things. Not really the place for it but I don't have anywhere else to put it.
Search term: https://www.google.com/search?q={searchTerms} vr site:steampowered.com&btnI

Will search for a VR game on the steam site only using the "I'm Feeling Lucky" feature so will take the top search result. 
The "&btnI" is what causes the lucky part.
Occasionally it will not got to the top result but display a search page. I think it does this when The Googs is not sure they have it right but I cannot say for sure and that may be giving them too much credit. 😃

**Comment by mrshanepaul** [comment]
Absolutely. Otherwise people miss out on the utility of the tool.

The OS world is so very full of badly documented, basically hidden features that almost no one uses....

**Comment by CanisLupus** [comment]
True. Like the Ctrl+Win+C shortcut on Windows 10, which I find very amusing. ;D

**Comment by ZIEXED** [comment]
I don't want to bother you but, I think this is easy to implement, any progress here?

**Comment by CanisLupus** [comment]
Hey ZIEXED, it's no problem. :) I went and looked at this today after your mention, since I hadn't yet and this is an important issue, but it's going to take a while more. Deciding how to do the parsing is especially lengthy.

I wanted to allow replacements of strings of characters instead of only single characters (so you could write (something,nope) and replace all "something"s with "nope"s in whatever you selected as the search). There is also the issue of characters I assume as separators (parenthesis, comma, etc) also being able to appear in the strings themselves, which complicates parsing. Regex was not looking good. :p

Daniel

**Comment by CanisLupus** [comment]
Just to let everyone know, I've made progress with this today since I had some time, but it's still gonna be a while because now I'm going away until the end of next week. Until then! :)

Daniel

**Comment by CanisLupus** [comment]
So... I had a kind of hiatus of almost 2 months in regards to SSS development... I apologize to those that were awaiting this feature to solve problems with some sites. :(

I've returned today and finally finished and uploaded it (v3.23.0) along with other minor pending changes. It's implemented similarly to my previous suggestion, but I changed a few things, namely removing the ":" and using "{", "|", and "}" instead of "(", "," and ")".

Examples:
- Using {searchTerms{ |+}} replaces spaces with + symbols.
- Using {searchTerms{ |+}{SSS|something else}} replaces spaces with + symbols and "SSS" with "something else" (yes, it supports text, not only single characters).
- Since SSS uses them to delimit the replacement pairs, to use special characters \, {, |, or } in the replacements you should precede them with a \. For example, {searchTerms{\{|\}}} (confusing, I know!) replaces "{" with "}" in your text and {searchTerms{ |\\} replaces spaces with a "\".

You can now also use several {searchTerms} in the same search engine URL if you wish, each with different substitutions even. I haven't found a use for this but I'm sure it exists for some weird website. ;)

I'm leaving this issue open until feedback comes in. It might need some changes, though I hope not! ;)

Cheers,
Daniel

**Comment by magOwl** [comment]
Hi,
It seems I've encountered a slight problem after this update (although I really appreciate the added functionality). I frequently search over a line break, a `<br>` in this specific case, which causes the last word before the linebreak and the first word after to be joined in the search, which consequently messes up the search. 

Like this:

**Some text**
**Additional text**

and the search string becomes **Some textadditional text**.

Is there a way for me to replace/substitute the linebreak so we get the space back between the words using the new implemented substitution method, or is it something that has to be made on your end?

I could probably whip something up with AHK though, but I thought I would ask, as well as let you know in any case.

Thanks for the great plugin btw, been using it more or less since the day it showed up!

Cheers!

**Comment by magOwl** [comment]
Hey!

Yeah I figured :)

I'm pretty sure this didn't occur before this update (otherwise I wouldn't have noticed it now). But you guys talked about stripping extra spaces and whatnot earlier in this thread, maybe that has something to do with it? As an example, the built in context search in FF works like that (adding the space on a linebreak) as well as some other plugs like yours I've tried (and as I said, I'm 99.9% sure yours did as well before the update).

I have no idea if using AHK would work, but it's very good at solving little things like this just for personal use, so I'm pretty sure it could be done. It has some nice hooks and can inject itself in many places and contexts etc. I have a small script that's always running that changes everything from mouse behavior in vlc (which couldn't be changed natively for some reason) to some simple autocompletes. It's pretty useful for random little things that annoy you :)

But I hope you figure something out and I'll gladly give some feedback if I can.

Cheers!

**Comment by CanisLupus** [comment]
Oops sorry, it seems I don't even know my own extension! :) You're completely right, from v3.22.0 to v3.23.0 both of the issues you report (line break and &) started happening (although previously it didn't deal with line breaks either, but they were sent in the text and the search engine dealt with them as it wished).

Good point about the stripping, though it should only affect the start and end of the selected text.

Seems interesting, maybe I'll have to give AHK a more detailed look. Where I work I used it to automate a VERY cumbersome process of uploading achievement metadata to Steam for our games. It simulated clicking file browser dialogs to upload icons and "tabbed" through dozens of input fields while writing the names and descriptions for the achievements. Worked pretty well, but it was massively hack-ish. :)

Thank you a lot for the reports. I probably wouldn't have noticed this in a long while.  I'll look into this ASAP. I'll let you know.

Cheers!

**Comment by CanisLupus** [comment]
Well, I had forgotten to add the call to the "encodeURIComponent" function, which deals with weird characters like "&", which are unsupported in URL arguments. :) It was there before, but along with the changes to support the character replacements I forgot to add it back. Oops!

That actually fixed the line break issue as well (back to how it was), but I decided that it's better to just replace any line breaks with spaces anyway because some engines might not deal with the encoded newline, so I also did that.

So it should be fine in SSS v3.24.1.

Thanks again for the reports! :)

**Comment by magOwl** [comment]
Yeah AHK is a bit hackish, but it works well most of the time :)

Ok cool, quick turnaround! Good job man! Looking forward to the update.

No worries, cheers!

**Comment by CanisLupus** [comment]
Thank you! :)

**Comment by CanisLupus** [comment]
Hey everyone, I'm closing this issue but if you find some problem directly related to it feel free to reopen!

Cheers! :)
Daniel

**Comment by xbkingx** [comment]
Sorry to re-open old wounds, er... issues, but I;m having some trouble related to spaces in search terms and substitution.

tl;dr - Replacing spaces with nothing (deleting spaces) doesn't seem to work. Replacing spaces with %20 in the SearchTerms gets around the problem.

I added several (legit) phone number lookup sites for my parents that get a ton of robocalls. The phone numbers in their VOIP provider's call log displays numbers as: `(###)<space>###-####` 

The phone lookup sites parse searches as: `site/phone/###/###/####/`

I can delete the prefix "(", sub the ")", and delete "-". I can't seen to sub ")<space>" or delete the space after ")".

I've tried `{searchTerms{(|}{)|-}{ |}}` and `{searchTerms{(|}{)|-}{ |}{) |-}}`

I tried to fall back to replacing the space instead of ")" with "/", but it's not recognized. This doesn't work: `{searchTerms{(|}{ |/}{-|/}}`, and results in `###)<space>###/####'. Same for `{searchTerms{(|}{)|/}{-|/}}`

Something about spaces by punctuation or replacing with "/" appears to be wonky. 

Solution: 

I finally got it to work by using %20 in place of space, like `{searchTerms{(|}{)|/}{%20|}{-|/}}`. So, something happens during one of the encoding hand-offs that loses the token, like the number of non-whitespace characters are counted followed by checking the string for all characters in the token ("I'm checking for '`<space>` + )' with a strong length of 1."). Just speculating.

**Comment by CanisLupus** [comment]
Absolutely no problem. :) You found something that was not working as expected, and I thank you for the detailed report and examples. You are right, and the worst is that even my own example in the Instructions section (advanced usage) doesn't work because it has a space substitution.

The issue seems to be calling encodeURIComponent, which encodes characters to be valid in the URL. Of course I completely forgot that calling it before the substitutions happen will break stuff like this. -.- Sigh...  :p The solution would be to simply encode the search AFTER the text replacements, but I'll have to see if that causes other problems first.

I'm glad you found out how to defeat the system, though. ;) When this gets fixed I'll mention it here.

I'm re-opening this issue until I fix this.

**Comment by xbkingx** [comment]
Whew! I'm not crazy (at least in this regard). I thought I was doing something wrong because the example wasn't working. 

Absolutely awesome extension, btw. It's the first thing I install (er, okay, after uBlock) on every new FFx install.

**Comment by CanisLupus** [comment]
Oops, I'm sorry for not replying to your last message, @xbkingx. Completely forgot. :( You were definitely not crazy. ;) And thanks for sticking with SSS!

I just submitted version 3.28.0, which will hopefully behave as expected regarding replacements, since the URL encoding is now only done after the replacements, as it always should have. Also my apologies for the long time this fix took; yesterday and today I finally had some decent time for SSS.

Please let me know if anything is still not working right. :)

Cheers!
Daniel

**Comment by xbkingx** [comment]
Sorry for my lag in response, but just got a chance to test the changes annnnnnd... it mostly works, but the correction caused another bug, probably in the re-encoding on the way out. Or this is a separate bug, so let me know if I should start a new issue.

Before the latest update - using ASCII codes worked in place of special, non-reserved characters (`%20` in place of `<space>`) on the source string. (I think) special, non-reserved characters worked in the destination string. So, IIRC, `{%20|/}` worked, but `{ |/}` did not.

After the update - the problem is flipped, so that special, non-reserved characters work in the source, but not in the destination. The re-encoding leaves the destination in ASCII format. So, `{ |}` works, but `{ |/}` results in `%2F` in the URL. Also, replacing `/` in the destination with `%2F` ( `{ |%2F}` ) goes a little bonkers and results in `%252F` being added.

The good news is that it still works for my use case, but might not for everyone. If you want to test these further, here is what I'm looking at:

```
https://www.nomREMOVEorobo.cREMOVEom/lookup/{searchTerms{(|}{)|-}{ |}}
--- searchTerm format should be: /###-###-####

https://www.numberREMOVEguru.cREMOVEom/phone/{searchTerms{(|}{)|/}{ |}{-|/}}/
--- searchTerm format should be: /###/###/####

https://whocREMOVEalled.uREMOVEs/lookup/{searchTerms{ |}{-|}{(|}{)|}}
--- searchTerm format should be: /##########


Using these as tests:
111-111-1111
(111)111-1111
(111) 111-1111
111)111-1111
111) 111-1111
(plus all above with a trailing space)

```

Again, no big deal since the second site recognizes %2F as /, but thought I'd mention it in case others report problems.

Love love love the extension. Keep up the great work.

**Comment by CanisLupus** [comment]
No problem, I was almost a month late. :p

Thanks for the detailed report, but I got a little confused by the EDIT! :) Let's see if we can get on the same page.

Just to be 100% clear: the encoding to a URL-compatible format has to be done at some point in the process, otherwise some queries will effectively break if they have characters with special meaning for URLs. Spaces are fine for most websites, but the specification still says they get encoded to %20.

I was reading everything you wrote and nodding along, since you describe what I expected the behaviour to be. :) The process for SSS is:

1. selected text gets grabbed as-is
2. any text replacements are applied, in order
3. URL encoding is applied over the result

Previously, the process was broken since steps 2 and 3 were switched, so there were never any spaces to replace, because all of them would be %20.

Now that it's working properly, you should never use escape sequences such as %20 or %2F in the replacements (well, unless you really want to replace that in text from webpage, of course) because they now literally mean "character `%`, followed by `2`, followed by `0`". I was probably not clear about this. No one should have to use escaped sequences; that was just a result of my dumb mistake heheh.

`{ |/}` resulting in `%2F` in the URL is normal because the space was replaced by the slash, but a slash in a URL must be encoded as `%2F`.

What previously worked, such as `{%20|/}`, now won't since now `%20` it's just 3 chars, not a space. Likewise, imagine the replacement `{/|%2F}` on the example `/---/`, selected on a page:

1. SSS grabs `/---/`
2. replacement `{/|%2F}` changes it to `%2F---%2F`
3. URL gets encoded and, since `%` itself is a special character (encoded as `%25`), the string becomes the extremely ugly `%252F---%252F`

Sites will interpret the URL and determine that `%252F---%252F` means `%2F---%2F`, which is expected since `%2F` doesn't mean anything special at this point; it's just 3 chars. 

The EDIT left me a bit confused. Is there a website that is not working when using non-escaped replacements? Can you tell me more? I may not be seeing a consequence of this, but it seemed correct in my head until now.

Cheers!
Daniel

**Comment by xbkingx** [comment]
Yeah, sorry about the drive by edit. Turns out it was all my bad.

The second site changed their query format since last I checked, so (I think) the non-numeric characters are ignored and / is no longer used to separate phone number segments (probably drove their programmers mad with the old format).

While updating my old search strings to work with the new version I apparently missed a couple that had been tweaked to bypass the old bug by replacing space with `%20`. 

Seeing the ignored `%2F` from my first mistake and the `%20` from the second mistake in the URL led me to believe there were possibly a couple swaps between URL formats internally and one was skipped.

But, no, my bad. Umm... carry on... and thanks again.

**Comment by CanisLupus** [comment]
Aaaah all is clear now. No problem! Glad to know that everything is working. ;)

Cheers!

**Comment by LegitLlama** [comment]
@CanisLupus 

Hi!

This URL encoding thing has one drawback: I cannot search for URLs "as is", which seems necessary for a website like the Wayback Machine.

To reproduce, you can use ```https://web.archive.org/web/*/{searchTerms}``` which expects an URL like ```https://github.com/``` as search terms. You'll see that the site does not support the encoded URL ```https%3A%2F%2Fgithub.com%2F```, and really requires the ```://```. Unfortunately Firefox itself behaves like your add-on with the built-in keyword URL feature ```https://web.archive.org/web/*/%s```. (I understand it's supposed to be the standard way to do this)

Any workaround to this, whether in Firefox or SSS ? Am I doomed to never be able to do one click searches on the Wayback Machine ?

Thanks either way :)

&nbsp;



Note: Searching directly for ```https://web.archive.org/web/*/https://github.com/``` in Firefox's URL bar does work.

**Comment by CanisLupus** [comment]
Hmm I was under the impression that I'd already seen a problem with the Wayback Machine and sure enough, it was issue #183, but in that case it was because the user was using the {href} property in `https://web.archive.org/web/*/{href}` and it was being (wrongly) encoded.

For actually SELECTING a link and searching for that using {searchTerms} it will always be encoded, unfortunately. Can you please create another issue for adding a new option to search engines that disables their forced URL encoding?

Thanks for reporting this!

**Comment by LegitLlama** [comment]
Neat, I didn't notice that ```{href}``` worked, though it's limited to the URL of the page that's being clicked. Being able to select text that happens to be a URL, or to input text in SSS's pop-up text field, and then searching for them on the Wayback Machine is indeed a slightly different use case. I'll make an issue then, thank you for being so responsive!

&nbsp;

What about searching for the URL of \*a link\* ? For instance, make that it is possible from [this link](https://github.com/) to search for ```https://github.com/``` on e.g. Wayback Machine. Maybe that's out of the scope for this add-on though, and it looks like a feature that needs some proper thinking as well, so I'm not suggesting it as much as wondering what you would think about it.

**Comment by CanisLupus** [comment]
I did have requests for link-related search in #120 and #179, but these were for the text of a link, not the URL of the link, so yours is a bit different. :) SSS has the code for at least #179 but it is currently disabled until I find an acceptable way to solve an inconsistency that is a bit hard to explain here.

It would be a bit hard to somehow allow searching for the text or URL of a link, but I can think about it when I return to this topic.

**Comment by LegitLlama** [comment]
Yeah this feature sounds tricky to implement in a neat way that's on the level of the rest of the add-on.

Unrelated but might be food for thought if you ever come back to this: Firefox allows to select the text of a link without risk of "browsing away" if the user holds the ALT key as they select the text of the link. Not sure how this can fit in smoothly with an hypothetical ability of SSS to distinguish intent to search for URL rather than text. Just saying it exists.

Cheers!
